### PR TITLE
Ability to ignore some controller by Kind

### PR DESCRIPTION
**Pull request by k11h-de**
This PR fixes https://github.com/FairwindsOps/goldilocks/issues/534 

## Checklist
* [x] I have signed the CLA
* [x] I have updated/added any relevant documentation

### What changes did you make?
adding a new arg `--ignore-controller-kind=Job,CronJob` that prevents certain controller kinds from being considered

## notes
I am sorry that I am not following your contribution guidelines. I am not a developer and was using GitHub Copilot to make these changes. The `make tidy` and `make build` are compiling without any errors - I have only limited resources and permissions on my works laptop. I would be grateful if you would still consider this improvement.

### CODE REVIEW

File: pkg/vpa/vpa.go [hunk 1]
```diff
@@ -157,14 +158,19 @@ func (r Reconciler) namespaceIsManaged(namespace *corev1.Namespace) bool {
...
 	defaultMinReplicas, _ := vpaMinReplicasForResource(ns)
 
 	// these keys will eventually contain the leftover vpas that do not have a matching controller associated
 	vpaHasAssociatedController := map[string]bool{}
 	for _, controller := range controllers {
+		// Check if the controller kind is in the ignore list
+		if contains(ignoreControllerKind, controller.Kind) {
+			continue
+		}
+
```
Comment (sudermanjr) [inline pkg/vpa/vpa.go:173]: We would prefer to use https://github.com/samber/lo for this type of thing rather than writing one-off helper functions.

**Review by sudermanjr — CHANGES_REQUESTED** [review]
Thanks for the PR.

Adding a new flag will require end-to-end tests for this flag at a minimum.

File: pkg/vpa/vpa.go (hunk 1 above)
Comment (k11h-de) [reply pkg/vpa/vpa.go]: Thanks for your inout on how to improve the PR.
Could you please review the end-to-end test and change to use "github.com/samber/lo" instead the one-off helper function?

Thanks a lot

Comment (k11h-de) [reply pkg/vpa/vpa.go]: @sudermanjr 
Sorry, I am not so familiar with the work flows (and buttons) in Github - didn't meant to spam you.

File: pkg/vpa/vpa.go [hunk 2]
```diff
@@ -51,6 +52,7 @@ type Reconciler struct {
 	DryRun                bool
 	IncludeNamespaces     []string
 	ExcludeNamespaces     []string
+	ignoreControllerKind  []string
```
Comment (sudermanjr) [inline pkg/vpa/vpa.go:55]: This field should be exported to match the other fields in this struct. This way it can be used by other packages

```suggestion
	IgnoreControllerKind  []string
```

File: cmd/controller.go [hunk 3]
```diff
@@ -37,6 +38,7 @@ func init() {
 	controllerCmd.PersistentFlags().BoolVar(&dryRun, "dry-run", false, "If true, don't mutate resources, just list what would have been created.")
 	controllerCmd.PersistentFlags().StringSliceVar(&includeNamespaces, "include-namespaces", []string{}, "Comma delimited list of namespaces to include from recommendations.")
 	controllerCmd.PersistentFlags().StringSliceVar(&excludeNamespaces, "exclude-namespaces", []string{}, "Comma delimited list of namespaces to exclude from recommendations.")
+	controllerCmd.PersistentFlags().StringSliceVar(&ignoreControllerKind, "ignore-controller-kind", []string{}, "Comma delimited list of controller kinds to exclude from recommendations.")
```
Comment (sudermanjr) [inline cmd/controller.go:41]: Below this, around line 52, this should be added to the vpaReconciler

File: pkg/vpa/vpa.go [hunk 4]
```diff
@@ -110,7 +112,7 @@ func (r Reconciler) ReconcileNamespace(namespace *corev1.Namespace) error {
 		return err
 	}
 
-	return r.reconcileControllersAndVPAs(namespace, vpas, controllers)
+	return r.reconcileControllersAndVPAs(namespace, vpas, controllers, r.ignoreControllerKind)
```
Comment (sudermanjr) [inline pkg/vpa/vpa.go:115]: This field is already contained in the reconciler, no need to pass it like this.

```suggestion
	return r.reconcileControllersAndVPAs(namespace, vpas, controllers)
```

File: pkg/vpa/vpa.go [hunk 5]
```diff
@@ -157,14 +159,19 @@ func (r Reconciler) namespaceIsManaged(namespace *corev1.Namespace) bool {
 	return r.OnByDefault
 }
 
-func (r Reconciler) reconcileControllersAndVPAs(ns *corev1.Namespace, vpas []vpav1.VerticalPodAutoscaler, controllers []Controller) error {
+func (r Reconciler) reconcileControllersAndVPAs(ns *corev1.Namespace, vpas []vpav1.VerticalPodAutoscaler, controllers []Controller, ignoreControllerKind []string) error {
```
Comment (sudermanjr) [inline pkg/vpa/vpa.go:162]: Same as above, should be in the Reconciler

```suggestion
func (r Reconciler) reconcileControllersAndVPAs(ns *corev1.Namespace, vpas []vpav1.VerticalPodAutoscaler, controllers []Controller) error {
```

File: pkg/vpa/vpa.go [hunk 6]
```diff
@@ -157,14 +159,19 @@ func (r Reconciler) namespaceIsManaged(namespace *corev1.Namespace) bool {
...
 	defaultUpdateMode, _ := vpaUpdateModeForResource(ns)
 	defaultResourcePolicy, _ := vpaResourcePolicyForResource(ns)
 	defaultMinReplicas, _ := vpaMinReplicasForResource(ns)
 
 	// these keys will eventually contain the leftover vpas that do not have a matching controller associated
 	vpaHasAssociatedController := map[string]bool{}
 	for _, controller := range controllers {
+		// Check if the controller kind is in the ignore list
+		if lo.Contains(ignoreControllerKind, controller.Kind) {
```
Comment (sudermanjr) [inline pkg/vpa/vpa.go:171]:
```suggestion
		if lo.Contains(r.IgnoreControllerKind, controller.Kind) {
```

**Comment by k11h-de** [comment]
Hi Andy,

Thanks a lot for the suggestions. I merged them.
Please let me know if you wish any other changes.

Sorry again for my poor coding skills.
Cheers and warm wishes from Berlin
Karsten

File: cmd/controller.go (hunk 3 above)
Comment (sudermanjr) [reply cmd/controller.go:41]: I don't believe this has been added to the actual instance of the struct that gets declared on lie 49

**Review by sudermanjr — CHANGES_REQUESTED** [review]

File: cmd/controller.go (hunk 3 above)
Comment (k11h-de) [reply cmd/controller.go:41]: You're right. I made the requested change.

**Review by sudermanjr — APPROVED** [review]
Awesome. Thanks!
